package/MDAnalysis/analysis/msd.py

```diff
@@ -0,0 +1,213 @@
+class MeanSquaredDisplacement(object):
+    r"""Class representing a density on a regular cartesian grid.
+
+    Parameters
+    ----------∏
+    u : 
+        An MDAnalysis Universe :class:`Universe`
+    selection : 
+        An MDAnalysis selection string
+    
+    
+
+    Attributes
+    ----------
+
+
+
+
+    Notes
+    -----
+    Notes
+
+
+    See Also
+    --------
+   
+
+    Examples
+    --------
+    Typical use:
+
+    
+
+    """
+
+    def __init__(self, u, selection, msd_type='xyz', fft=True):
+
+        #args
+        self.u = u
+        self.selection = selection
+        self.msd_type = msd_type
+        self.fft = fft
+        
+        #local
+        self.dim_fac = 0
+        self._dim = None
+        self.atoms = None
+        self.n_frames = len(self.u.trajectory)
+        self._position_array = None
+
+        #result
+        self.timeseries = None
+        #prep
+        self._prepare()
+
+            
+    def _prepare(self):
+        self.parse_msd_type()
+        self.select_reference_positions()
+        
+    def parse_msd_type(self):
+
+        if self.msd_type == 'xyz': # full 3d
+            self._dim = [0,1,2]
+            self.dim_fac = 3.0
+            
+        elif self.msd_type == 'xy': # xy
+            self._dim = [0,1]
+            self.dim_fac = 2.0
+
+        elif self.msd_type == 'xz': # xz
+            self._dim = [0,2]
+            self.dim_fac = 2.0
+
+        elif self.msd_type == 'yz': # yz
+            self._dim = [1,2]
+            self.dim_fac = 2.0
+
+        elif self.msd_type == 'x': # x
+            self._dim = [0]
+            self.dim_fac = 1.0
+
+        elif self.msd_type == 'y': # y
+            self._dim = [1]
+            self.dim_fac = 1.0
+
+        elif self.msd_type == 'z': # z
+            self._dim = [2]
+            self.dim_fac = 1.0
+
+        else:
+            raise ValueError('invalid msd_type specified')
+
+    def select_reference_positions(self):
+        self._position_array = self.u.trajectory.timeseries(self.u.select_atoms(self.selection),order='fac') 
+        self.N_particles = self._position_array.shape[1] 
+    
+    def run(self):
+        if self.fft == True:
+            self.timeseries = self._run_fft()
+        else:
+            self.timeseries = self._run_simple()
+
+    def _run_simple(self): # naieve algorithm without FFT
+        # _position_array is shape time, nparticles, 3
+        msds_byparticle = np.zeros([self.n_frames, self.N_particles])
+        lagtimes = np.arange(1,self.n_frames)
+        msds_byparticle[0,:] = np.zeros(self.N_particles) # preset the zero lagtime so we dont have to iterate through
+        for n in range(self.N_particles):
+            for lag in lagtimes:
+                disp = self._position_array[:-lag,n,self._dim if lag else None] - self._position_array[lag:,n,self._dim]
+                sqdist = np.square(disp, dtype=np.float64).sum(axis=1, dtype=np.float64) #accumulation in anything other than f64 is innacurate
```

Comment:
To add to the AnalysisBase discussion -- isn't this iterating over the frames? I think you could put this in `_single_frame()`.

Reply:
@lilyminium  I tried to avoid this as there is no return value associated with a single frame. if that makes sense

Reply:
Ah right, I see that it needs the whole position array.